graphql/writer.go

```diff
@@ -0,0 +1,112 @@
+package graphql
+
+import (
+	"encoding/json"
+	"sync"
+)
+
+// errorRecorder is a concurrency-safe way where we can record the first error
+// we get from executing the graphql query.
+type errorRecorder struct {
+	sync.Once
+	err error
+}
+
+func (e *errorRecorder) record(err error) {
+	if err == nil {
+		return
+	}
+	e.Do(func() {
```

Comment:
is this cheaper or more efficient than a sync.Mutex? would prefer mutex because i find it simpler

Reply:
Under the hood a sync.Once is a boolean atomic and a sync.Mutex.  If we did have a mutex we'd probably want to have the same functionality with a boolean atomic to quick-exit if the error had already been set (which would be more efficient than just a mutex.